packages/frontend/src/components/UnderlyingData/UnderlyingDataProvider.tsx

```diff
@@ -0,0 +1,274 @@
+    const closeModal = useCallback(() => {
+        resetQueryResults();
+    }, [resetQueryResults]);
+
+    useEffect(() => {
+        if (
+            state.shouldFetchResults &&
+            state.unsavedChartVersion.metricQuery.dimensions.length > 0
+        ) {
+            mutate();
+            setState({ ...state, shouldFetchResults: false });
+        }
+    }, [mutate, state, setState]);
+
+    const allFields = useMemo(
+        () => (explore ? getFields(explore) : []),
+        [explore],
+    );
+    const allDimensions = useMemo(
+        () => (explore ? getDimensions(explore) : []),
+        [explore],
+    );
+
+    const viewData = useCallback(
+        (
+            value: ResultRow[0]['value'],
+            meta: TableColumn['meta'],
+            row: ResultRow,
+        ) => {
+            if (meta?.item === undefined) return;
+
+            // We include tables from all fields that appear on the SQL query (aka tables from all columns in results)
+            const rowFields = Object.keys(row);
+            const fieldsInQuery = allFields.filter((field) =>
+                rowFields.includes(getFieldId(field)),
+            );
+            const tablesInQuery = new Set([
+                ...fieldsInQuery.map((field) => field.table),
+                tableName,
+            ]);
+
+            const availableDimensions = allDimensions.filter(
+                (dimension) =>
+                    tablesInQuery.has(dimension.table) &&
+                    !dimension.timeInterval &&
+                    !dimension.hidden,
+            );
+
+            // If we are viewing data from a metric or a table calculation, we filter using all existing dimensions in the table
+            const dimensionFilters =
+                !isField(meta?.item) || isMetric(meta?.item)
+                    ? Object.entries(row).reduce((acc, r) => {
+                          const [
+                              key,
+                              {
+                                  value: { raw },
+                              },
+                          ] = r;
+
+                          const dimensionFilter: FilterRule = {
+                              id: uuidv4(),
+                              target: {
+                                  fieldId: key,
+                              },
+                              operator: FilterOperator.EQUALS,
+                              values: [raw],
+                          };
+                          const isDimension = availableDimensions.find(
+                              (dimension) => getFieldId(dimension) === key,
+                          );
+
+                          if (
+                              isDimension ||
+                              isDateDimension(availableDimensions, key)
+                          ) {
+                              return [...acc, dimensionFilter];
+                          }
```

Comment:
We could use the `allDimensions` array to avoid the extra function/logic in `isDateDimension`

`availableDimensions` can be used just when deciding what dims to be included in the query in line 223